lua_cjson.c

```diff
@@ -713,22 +714,57 @@ static void json_append_data(lua_State *l, json_config_t *cfg,
     case LUA_TTABLE:
         current_depth++;
         json_check_encode_depth(l, cfg, current_depth, json);
-        len = lua_array_length(l, cfg, json);
-        if (len > 0 || (len == 0 && !cfg->encode_empty_table_as_object))
+
+        int as_array = 0;
+        int has_metatable = lua_getmetatable(l, -1);
+
+        if (has_metatable) {
+            lua_pushlightuserdata(l, &json_array);
+            lua_rawget(l, LUA_REGISTRYINDEX);
+            as_array = lua_rawequal(l, -1, -2);
+            lua_pop(l, 1); /* keep metatable */
+        }
+
+        if (as_array) {
+            lua_pop(l, 1); /* pop metatable */
+            len = lua_objlen(l, -1);
             json_append_array(l, cfg, current_depth, json, len);
-        else {
-            int as_array = 0;
-            if (lua_getmetatable(l, -1)) {
-                lua_pushlightuserdata(l, &json_empty_array);
-                lua_rawget(l, LUA_REGISTRYINDEX);
-                as_array = lua_rawequal(l, -1, -2);
-                lua_pop(l, 2);
+        } else {
+            if (has_metatable) {
+                /* swap metatable and table: table on top */
+                lua_pushvalue(l, -2);
+                lua_pushvalue(l, -2);
+                lua_replace(l, -4);
+                lua_replace(l, -2);
             }
 
-            if (as_array) {
-                json_append_array(l, cfg, current_depth, json, 0);
+            len = lua_array_length(l, cfg, json);
+
+            if (len > 0 || (len == 0 && !cfg->encode_empty_table_as_object)) {
+                if (has_metatable) {
+                    lua_remove(l, -2); /* remove metatable */
+                }
+
+                json_append_array(l, cfg, current_depth, json, len);
             } else {
-                json_append_object(l, cfg, current_depth, json);
+                if (has_metatable) {
+                    /* swap metatable and table: metatable on top */
+                    lua_pushvalue(l, -2);
+                    lua_pushvalue(l, -2);
+                    lua_replace(l, -4);
+                    lua_replace(l, -2);
```

Comment:
We can avoid those swapping again (and the `lua_remove()` above) if we call `lua_getmetatable()` a second time here instead.